rdr_service/api/ppsc_intake_api.py

```diff
@@ -0,0 +1,123 @@
+        required_keys = ['activity', 'eventType', 'participantId', 'dataElements']
+
+        # Check required keys in payload
+        if all([key in req_data for key in required_keys]) \
+                and all([val for val in req_data.values() if val is not None]):
+            pass
+        else:
+            raise BadRequest(f'Invalid Intake API Payload: Required keys: {required_keys}')
+
+        # Check Activity is valid
+        if req_data['activity'] not in self.intake_activities:
+            raise BadRequest(f'Invalid Intake API Payload: Invalid Activity: {req_data["activity"]}')
+
+        event_type_lookup_str = 'ppsc_intake_' + req_data['activity'].lower().replace(' ', '_') + '_event_types'
+
+        # Check Event Type is valid
+        if req_data['eventType'] not in config.getSettingJson(event_type_lookup_str):
+            raise BadRequest(f'Invalid Intake API Payload: Invalid EventType: {req_data["eventType"]}')
+
+        # Check for Event Authored Date
+        self.activity_date_time_value = next((item['dataElementValue'] for item in req_data['dataElements'] if
+                                              item['dataElementName'] == 'activity_date_time'), None)
+
+        # Check if the activity_date_time_value is not None
+        if self.activity_date_time_value is not None:
+            try:
+                # Parse the datetime string to a datetime object using dateutil.parser
+                parsed_datetime = parser.isoparse(self.activity_date_time_value)
+
+                # Check if the datetime is in UTC
+                if parsed_datetime.tzinfo is not None and parsed_datetime.tzinfo.utcoffset(parsed_datetime) is not None:
+                    self.activity_date_time_value = parsed_datetime
+                else:
+                    raise BadRequest("The activity_date_time_value is missing timezone info or not UTC.")
+
+            except ValueError:
+                raise BadRequest("The activity_date_time_value is not valid.")
+        else:
+            raise BadRequest("No activity_date_time_value provided.")
+
+    def handle_event_insert(self, *, req_data: dict) -> dict:
+        activity_record = list(filter(lambda x: x.name.lower() == req_data['activity'].lower(),
+                                      self.activity_records))[0]
+
+        # Insert participant_event_activity record
+        participant_event_activity_dict = {
+            'activity_id': activity_record.id,
+            'participant_id': self.dao.extract_prefix_from_val(req_data['participantId']),
+            'resource': req_data
+        }
+
+        # Validate participant ID and insert
+        try:
+            participant_event_activity = self.participant_event_activity_dao.insert(
+                self.participant_event_activity_dao.model_type(**participant_event_activity_dict)
+            )
+        except IntegrityError:
+            raise NotFound(f"Participant with ID {req_data['participantId']} not found")
```

Comment:
Will this be thrown on the insert if the FK does not exist? Should we add this check to the validate_payload method so nothing gets thru to this point?

Reply:
For performance reasons, I was trying to avoid a second call to the database to lookup the PID. It may not be that big of a performance hit. This method here does feel a little messy. This is on the insert to `participant_event_activity`, where the FK does exist, so for now it should work. You suggestion is something I  could try on a future iteration when I revisit the validation.